design.md

```diff
@@ -0,0 +1,106 @@
+# Teleport assessment design document
+
+## Goal
+Implement a prototype job worker service that provides an API to run arbitrary
+Linux processes.
+
+## Architecture
+The system is made up of 3 components. The worker library, the grpc api and the client/cli interface to run commands
+* The worker library provides methods to start, stop and get the status of a running job.
+A method is also provided to stream the output of a job. The api can also be queried for all previous output of a running job. The server also provides a stream for clients to listen to job output.
+* The grpc api is a grpc service that sits on top of the worker library. It provides methods to start, stop and get the status of a running job remotely. 
+* The client/cli is an interface for users to start, stop and get the status of a running job on a remote server.
+
+## Worker library API
+The worker library consists of 2 components. The supervisor and the process.
+### Supervisor API
+The supervisor is responsible for managing all jobs in the system. It provides apis to start, stop, get the output and get the status of a job. 
+* `Newsupervisor()` - Returns a new instance of a supervisor
+* `Start(cmd string, args []string, opts...Options)` - Starts a process and return a unique job id. The job id is generated by the supervisor.
+* `Stop(id string)` - Stops a running process by sending the process group a `SIGTERM` signal.
+* `Run()` - Run the supervisor and listen for incoming job requests.
+* `WatchOutput(id string, c chan ProcessOutput)` - Add an output channel to subscribe to process output. All previous output, if any, from the process is replayed before sending new messages to the channel.
+* `UnWatchOutput(id string, c chan ProcessOutput)` - Unsubscribe to the process output.
+* `GetOutput(id string)` - Get all lines of output from the job's line buffer. Useful when getting the output of a finished job.
```

Comment:
How will this work? Just curious how "pre-existing" output would be returned to the client - streamed as well, or as a single "blob"?